app/models/cart.rb

```diff
@@ -1,14 +1,24 @@
 class Cart < ApplicationRecord
-  has_many :cart_products, dependent: :destroy
-  has_many :products, through: :cart_products
+  has_many :items, class_name: "CartProduct", dependent: :destroy
+  has_many :products, through: :items
 
   validates :session_key, presence: true
 
-  def self.with_cart_products
-    includes(:cart_products)
+  def self.with_items
+    includes(:items)
+  end
+
+  def update_item_quantity(product_id, quantity)
+    item = items.find_or_initialize_by(product_id: product_id)
+    item.quantity += quantity.to_i
```

Comment:
Why are you adding new quantity to the existing, instead of just replacing it with the new value?